### PR TITLE
DataGrid PDF Export - Print Info Calculator - Simple cases: table pos/width, cell coordinates

### PR DESCRIPTION
**Pull request by KuznetsovVN**

### CODE REVIEW

File: js/exporter/jspdf/export_print_info_calculator.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            jsPDFDocument,
+            margin,
+            startX,
+            startY,
+            cells,
+            tableWidth,
+            columnStyles
+        } = options;
+
+        margin.top = margin.top ?? margin.all;
```
Comment (novsstation) [inline js/exporter/jspdf/export_print_info_calculator.js]: Is the "all" option really needed?

**Review by novsstation — DISMISSED** [review]

**Review by EugeniyKiyashko — APPROVED** [review]

**Review by novsstation — APPROVED** [review]

File: js/exporter/jspdf/export_print_info_calculator.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            cells: []
+        };
+    },
+
+    calculate: function(options) {
+        options = extend(true, {}, this._getDefaultOptions(), options);
+
+        const {
+            jsPDFDocument,
+            margin,
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_print_info_calculator.js:25]: imho it's better to sync our API with jsPdf.text: 
`doc.text("Hello world!", 10, 10);`
http://raw.githack.com/MrRio/jsPDF/master/docs/jsPDF.html#text

The doc.text will be used to create header/footer/left/right notes and it doesn't provide 'margin' in root args.

File: js/exporter/jspdf/export_print_info_calculator.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            startX,
+            startY,
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_print_info_calculator.js:27]: http://raw.githack.com/MrRio/jsPDF/master/docs/jsPDF.html#text

doc.text calls these arguments 'X,Y' without the 'start' prefix

File: js/exporter/jspdf/export_print_info_calculator.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+    _calculateTableWidth: function(jsPDFDocument, leftIndent, rightIndent) {
+        const pageSize = jsPDFDocument.internal.pageSize;
+        const pageWidth = isDefined(pageSize.width) ? pageSize.width : pageSize.getWidth();
+        return pageWidth - (leftIndent + rightIndent);
+    },
+
+    _calculateColumnsStyles: function(tableWidth, columnStyles, cells) {
+        let columnCount = 0;
+        cells[0]?.forEach((cell, _) => {
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_print_info_calculator.js:57]: Why cell.colSpan is calculated from cells[0] only ? Is it Header cells? Or Data cells? How empty grid will be processed?

File: js/exporter/jspdf/export_print_info_calculator.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            if(isNumeric(columnStyles[i]?.width)) {
+                columnsWithAssignedWidth.push(i);
+                unAllocatedWidth -= columnStyles[i].width;
+            }
+        }
+
+        let columnAutoWidth = 0;
+        if(unAllocatedWidth > 0) {
+            columnAutoWidth = unAllocatedWidth / (columnCount - columnsWithAssignedWidth.length);
+        }
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_print_info_calculator.js:74]:
```
else {
    add an error message to console
}
```

File: js/exporter/jspdf/export_print_info_calculator.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            const getRowHeight = () => {
+                return rowHeight;
+            };
+
+            for(let colIndex = 0; colIndex < cells[rowIndex].length; colIndex++) {
+                const prevCellCoordinates = (colIndex > 0) ? cells[rowIndex][colIndex - 1].getCoordinates() : null;
+                const colStyles = columnStyles[colIndex];
+
+                const xPos = isDefined(prevCellCoordinates) ? prevCellCoordinates.x + prevCellCoordinates.width : x;
+                const yPos = lastRowYPos;
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_print_info_calculator.js:102]: xPos  -> cellLeft/cellX
yPos -> cellTop/cellY

File: js/exporter/jspdf/export_print_info_calculator.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+                columnStyles[i].minCellHeight = this._defaultMinCellHeight;
+            }
+            if(!isNumeric(columnStyles[i]?.width)) {
+                columnStyles[i].width = columnAutoWidth;
+            }
+        }
+    },
+
+    _calculateCellsCoordinates: function(x, y, cells, columnStyles) {
+        let lastRowYPos = y;
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_print_info_calculator.js:90]: lastRowYPos -> lastRowTop/lastRowY

File: js/exporter/jspdf/export_print_info_calculator.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        for(let rowIndex = 0; rowIndex < cells.length; rowIndex++) {
+            let rowHeight = 0;
+            const getRowHeight = () => {
+                return rowHeight;
+            };
+
+            for(let colIndex = 0; colIndex < cells[rowIndex].length; colIndex++) {
+                const prevCellCoordinates = (colIndex > 0) ? cells[rowIndex][colIndex - 1].getCoordinates() : null;
+                const colStyles = columnStyles[colIndex];
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_print_info_calculator.js:99]: colStyles -> currentColumnStyles

File: testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+        QUnit.module('Print Info Calculator', moduleConfig, () => {
+            QUnit.test('Empty grid', function(assert) {
+                const done = assert.async();
+
+                const options = {
+                    jsPDFDocument: this.jsPDFDocument,
+                    cells: []
+                };
+                const calculatedOptions = ExportPrintInfoCalculator.calculate(options);
+                assert.strictEqual(calculatedOptions.columnStyles[0], undefined, 'columnStyles[0]');
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js:19]: assert is not clear
may be it should be "assert.strictEqual(calculatedOptions.columnStyles.length, 0, 'columnStyles.length');" ?

File: js/exporter/jspdf/export_print_info_calculator.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        for(let i = 0; i < columnCount; i++) {
+            if(isNumeric(columnStyles[i]?.width)) {
+                columnsWithAssignedWidth.push(i);
+                unAllocatedWidth -= columnStyles[i].width;
+            }
+        }
+
+        let columnAutoWidth = 0;
+        if(unAllocatedWidth > 0) {
+            columnAutoWidth = unAllocatedWidth / (columnCount - columnsWithAssignedWidth.length);
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_print_info_calculator.js:73]: (columnCount - columnsWithAssignedWidth.length) - can it be '0'?

File: js/exporter/jspdf/export_print_info_calculator.js [hunk 11]
```diff
@@ -0,0 +1,121 @@
+import { isDefined, isNumeric } from '../../core/utils/type';
+import { extend } from '../../core/utils/extend';
+
+export const ExportPrintInfoCalculator = {
+    _defaultMinCellHeight: 10,
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_print_info_calculator.js:5]: 10 ? cm/mm/in/pt?

File: testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+                    cells: [
+                        [{ content: 'text_1_1' }, { content: 'text_1_2' }],
+                        [{ content: 'text_2_1' }, { content: 'text_2_2' }]
+                    ]
+                };
+                const calculatedOptions = ExportPrintInfoCalculator.calculate(options);
+
+                const columnWidth = calculatedOptions.tableWidth / 2;
+                const startX = calculatedOptions.startX;
+                const startY = calculatedOptions.startX;
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js:67]: calculatedOptions.startX -> calculatedOptions.startY ?

File: testing/tests/DevExpress.exporter/jspdfParts/jspdfTestHelper.js [hunk 13]
```diff
@@ -92,4 +92,34 @@ class JSPdfDataGridTestHelper {
...
+    constructor(jsPDFDocument) {
+        this.jsPDFDocument = jsPDFDocument;
+    }
+
+    checkColumnWidths(expectedColumnWidths, options) {
+        const columnStyles = options.columnStyles;
+        const actualWidths = [];
+
+        for(let i = 0; i < expectedColumnWidths.length; i++) {
+            const width = columnStyles[i] ? columnStyles[i].width : 0;
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/jspdfTestHelper.js:105]: For a separate TablePrinter (width === undefined/null) should mean 'auto'.
IMHO we should change this.
And, it' not clear which class should make this choice: calculator or printer?

File: testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+                    cells: []
+                };
+                const calculatedOptions = ExportPrintInfoCalculator.calculate(options);
+                assert.strictEqual(calculatedOptions.columnStyles[0], undefined, 'columnStyles[0]');
+
+                done();
+            });
+
+            QUnit.test('1 column, 1 rows', function(assert) {
+                const done = assert.async();
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js:25]: IHMO we should not use 'async()' just to make test safe if it doesn't wait for a promise.

File: testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+            QUnit.test('2 columns, 1 rows', function(assert) {
+                const done = assert.async();
+
+                const options = {
+                    jsPDFDocument: this.jsPDFDocument,
+                    cells: [[{ content: 'text_1' }, { content: 'text_2' }]]
+                };
+                const calculatedOptions = ExportPrintInfoCalculator.calculate(options);
+
+                const columnWidth = calculatedOptions.tableWidth / 2;
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js:46]: It's better to set explicit tableWidth in tests and add few additional tests to check how it is automatically calculated.

File: js/exporter/jspdf/export_print_info_calculator.js [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+        this._calculateCellsCoordinates(x, y, cells, columnStyles);
+
+        return extend(true, {}, options, {
+            startX: x,
+            startY: y,
+            tableWidth: width
+        });
+    },
+
+    _calculateTableWidth: function(jsPDFDocument, leftIndent, rightIndent) {
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_print_info_calculator.js:49]: I think we should postpone automatic tableWidth calculation and work without it for several sprints: it's easy to specify it explicitly and we can declare this value as required at the current stage.

File: js/exporter/jspdf/export_print_info_calculator.js [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+
+        for(let i = 0; i < columnCount; i++) {
+            if(!isDefined(columnStyles[i])) {
+                columnStyles[i] = {};
+            }
+            if(!isNumeric(columnStyles[i]?.minCellHeight)) {
+                columnStyles[i].minCellHeight = this._defaultMinCellHeight;
+            }
+            if(!isNumeric(columnStyles[i]?.width)) {
+                columnStyles[i].width = columnAutoWidth;
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_print_info_calculator.js:84]: I think we should postpone automatic columnWidth calculation and work without it for several sprints: it's easy to specify it explicitly and we can declare this value as required at the current stage.

File: js/exporter/jspdf/export_print_info_calculator.js [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+        if(unAllocatedWidth > 0) {
+            columnAutoWidth = unAllocatedWidth / (columnCount - columnsWithAssignedWidth.length);
+        }
+
+        for(let i = 0; i < columnCount; i++) {
+            if(!isDefined(columnStyles[i])) {
+                columnStyles[i] = {};
+            }
+            if(!isNumeric(columnStyles[i]?.minCellHeight)) {
+                columnStyles[i].minCellHeight = this._defaultMinCellHeight;
```
Comment (IgnatovDan) [inline js/exporter/jspdf/export_print_info_calculator.js:81]: I think we should postpone automatic rowHeight calculation and work without it for several sprints: it's easy to specify it explicitly and we can declare this value as required at the current stage.

File: testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js [hunk 19]
```diff
@@ -0,0 +1,324 @@
...
+                    startX: 50,
+                    startY: 20,
+                    tableWidth: 100,
+                    cells: [
+                        [{ content: 'text_1_1' }, { content: 'text_1_2' }],
+                        [{ content: 'text_2_1' }, { content: 'text_2_2' }]
+                    ],
+                    columnStyles: {
+                        0: { minCellHeight: 10 },
+                        1: { width: 70, minCellHeight: 20 }
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js:300]: There should be one property to declare height for entire row

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

File: js/exporter/jspdf/export_print_info_calculator.js (hunk 11 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/export_print_info_calculator.js:5]: This means the measurement unit, which the customer is defined in jsPDF instance.
As well as in the autoTable plugin, in our approach I will use constants.
In the future, this constant can be replaced by getFonSize() value.

File: js/exporter/jspdf/export_print_info_calculator.js (hunk 2 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/export_print_info_calculator.js:25]: Margin is a padding for the page. What you mean here?
Clarify please.

File: js/exporter/jspdf/export_print_info_calculator.js (hunk 3 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/export_print_info_calculator.js:27]: This variables are used for the starting position for the table. For customer's cases.

File: js/exporter/jspdf/export_print_info_calculator.js (hunk 4 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/export_print_info_calculator.js:57]: The total number of columns is determined by the number of cells in the first row. If the number of columns for each row is different, then this is already incorrect data. For empty grid the number of columns must be zero.

File: js/exporter/jspdf/export_print_info_calculator.js (hunk 10 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/export_print_info_calculator.js:73]: > can it be '0'?

maybe, we can add this check

File: js/exporter/jspdf/export_print_info_calculator.js (hunk 8 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/export_print_info_calculator.js:99]: yes, we can rename this variable

File: js/exporter/jspdf/export_print_info_calculator.js (hunk 6 above)
Comment (KuznetsovVN) [reply js/exporter/jspdf/export_print_info_calculator.js:102]: ok

File: testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js (hunk 9 above)
Comment (KuznetsovVN) [reply testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js:19]: 'columnStyles' is not an array and has no length property. I think we need to add the 'checkColumnWidths' method to the Helper object and check it there.

File: testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js (hunk 12 above)
Comment (KuznetsovVN) [reply testing/tests/DevExpress.exporter/jspdfParts/jspdf.printInfoCalculator.tests.js:67]: yes, there is a typo

**Pull request closed without merging**